### ISSUE TITLE
Prepare the H2 database

### ISSUE THREAD

**Issue by mislavmatijevic** [foivz/pop-app-backend#5]
# Preparation

This issue introduces the database access to the application.

## Description

Application needs a database in order to store users.
Consult [the **new** ER diagram](https://github.com/foivz/pop-app-backend/wiki/ER-Diagram#new-diagram-v1) prepared for the database. There, an SQL definition of the database is provided.

It has been decided to use sequential IDs for now.

The database needs to be completely changeable. The H2 database is going to be used as an in-memory alternative to an actual production database. When the web service is hosted in the production environment, it is most probably going to be using MySQL database. Of course, the table structure shall remain the same.

Any updates to the ER diagram shall reopen this issue, if needed.

Credentials for accessing this in-memory instance of H2 database are:
| Name | Password |
| ------ | ---------- |
| db | admin

## DoD

| Condition | Description |
| ------ | ---------- |
| Given | Application is hosted on the machine. |
| When | User opens `http://localhost:8080/h2-console` URL in a browser on the same machine. |
| Then | User sees the H2 interface, with specified _JDBC URL = `jdbc:h2:mem:`_ as in "in-memory", defined [here](http://www.h2database.com/html/features.html#in_memory_databases). User logs in with credentials and sees the H2 database management interface, where they can manage tables and data. Tables are filled with mock data.

### Events
```
INSERT INTO `mydb`.`events` (`name`, `date_created`, `is_active`) VALUES
  ('Event 1', NOW(), 1),
  ('Event 2', NOW(), 1);
```

### Stores
```
INSERT INTO `mydb`.`stores` (`events_id_event`, `store_name`, `balance`) VALUES
  (1, 'Store 1', 0),
  (1, 'Store 2', 0),
  (2, 'Store 3', 0);
```

### Roles
```
INSERT INTO `mydb`.`roles` (`role_name`) VALUES
  ('buyer'),
  ('seller'),
  ('admin');
```

### Users
```
INSERT INTO `mydb`.`users` (`roles_id_role`, `events_id_event`, `stores_id_store`, `name`, `surname`, `email`, `username`, `password_salt`, `password_hash`, `date_registered`, `balance`, `is_accepted`) VALUES
  (1, 1, 1, 'User 1', 'Surname 1', 'user1@example.com', 'user1', 'salt1', 'hash1', NOW(), 0, 0),
  (1, 1, 1, 'User 2', 'Surname 2', 'user2@example.com', 'user2', 'salt2', 'hash2', NOW(), 0, 0),
  (1, 1, 1, 'User 3', 'Surname 3', 'user3@example.com', 'user3', 'salt3', 'hash3', NOW(), 0, 0),
  (1, 1, 2, 'User 4', 'Surname 4', 'user4@example.com', 'user4', 'salt4', 'hash4', NOW(), 0, 0),
  (1, 1, 2, 'User 5', 'Surname 5', 'user5@example.com', 'user5', 'salt5', 'hash5', NOW(), 0, 0),
  (1, 1, 2, 'User 6', 'Surname 6', 'user6@example.com', 'user6', 'salt6', 'hash6', NOW(), 0, 0),
  (1, 1, 2, 'User 7', 'Surname 7', 'user7@example.com', 'user7', 'salt7', 'hash7', NOW(), 0, 0),
  (1, 1, 2, 'User 8', 'Surname 8', 'user8@example.com', 'user8', 'salt8', 'hash8', NOW(), 0, 0),
  (1, 2, 3, 'User 9', 'Surname 9', 'user9@example.com', 'user9', 'salt9', 'hash9', NOW(), 0, 0),
  (1, 2, 3, 'User 10', 'Surname 10', 'user10@example.com', 'user10', 'salt10', 'hash10', NOW(), 0, 0),
  (1, 2, 3, 'User 11', 'Surname 11', 'user11@example.com', 'user11', 'salt11', 'hash11', NOW(), 0, 0),
  (1, 2, 3, 'User 12', 'Surname 12', 'user12@example.com', 'user12', 'salt12', 'hash12', NOW(), 0, 0),
  (1, 2, 3, 'User 13', 'Surname 13', 'user13@example.com', 'user13', 'salt13', 'hash13', NOW(), 0, 0),
  (1, 2, 3, 'User 14', 'Surname 14', 'user14@example.com', 'user14', 'salt14', 'hash14', NOW(), 0, 0),
  (1, 2, 3, 'User 15', 'Surname 15', 'user15@example.com', 'user15', 'salt15', 'hash15', NOW(), 0, 0);
```

### Products
```
INSERT INTO `mydb`.`products` (`name`, `description`, `image`, `price`, `quantity`) VALUES
  ('Product 1', 'Description 1', 'image1.jpg', 10, 100),
  ('Product 2', 'Description 2', 'image2.jpg', 20, 200),
  ('Product 3', 'Description 3', 'image3.jpg', 30, 300),
  ('Product 4', 'Description 4', 'image4.jpg', 40, 400),
  ('Product 5', 'Description 5', 'image5.jpg', 50, 500),
  ('Product 6', 'Description 6', 'image6.jpg', 60, 600),
  ('Product 7', 'Description 7', 'image7.jpg', 70, 700),
  ('Product 8', 'Description 8', 'image8.jpg', 80, 800),
  ('Product 9', 'Description 9', 'image9.jpg', 90, 900),
  ('Product 10', 'Description 10', 'image10.jpg', 100, 1000),
  ('Product 11', 'Description 11', 'image11.jpg', 110, 1100),
  ('Product 12', 'Description 12', 'image12.jpg', 120, 1200),
  ('Product 13', 'Description 13', 'image13.jpg', 130, 1300),
  ('Product 14', 'Description 14', 'image14.jpg', 140, 1400),
  ('Product 15', 'Description 15', 'image15.jpg', 150, 1500),
  ('Product 16', 'Description 16', 'image16.jpg', 160, 1600),
  ('Product 17', 'Description 17', 'image17.jpg', 170, 1700),
  ('Product 18', 'Description 18', 'image18.jpg', 180, 1800),
  ('Product 19', 'Description 19', 'image19.jpg', 190, 1900),
  ('Product 20', 'Description 20', 'image20.jpg', 200, 2000);
```

### Packages
```
INSERT INTO `mydb`.`packages` (`name`, `description`, `image`, `discount_on_items`, `amount`) VALUES
  ('Package 1', 'Description 1', 'package1.jpg', 5, 100),
  ('Package 2', 'Description 2', 'package2.jpg', 10, 200),
  ('Package 3', 'Description 3', 'package3.jpg', 15, 300),
  ('Package 4', 'Description 4', 'package4.jpg', 20, 400),
  ('Package 5', 'Description 5', 'package5.jpg', 25, 500);
```

### packages_has_products
```
INSERT INTO `mydb`.`packages_has_products` (`packages_id_package`, `products_id_product`) VALUES
  (1, 1),
  (1, 2),
  (1, 3),
  (2, 4),
  (2, 5),
  (2, 6),
  (3, 7),
  (3, 8),
  (3, 9),
  (4, 10),
  (4, 11),
  (4, 12),
  (5, 13),
  (5, 14),
  (5, 15);
```

### Invoices
```
INSERT INTO `mydb`.`invoices` (`stores_id_store`, `users_id_user`, `code`, `date_issued`, `discount`, `total`) VALUES
  (1, 1, 'INV001', NOW(), 0, 150),
  (2, 2, 'INV002', NOW(), 10, 250),
  (3, 3, 'INV003', NOW(), 20, 350),
  (1, 4, 'INV004', NOW(), 0, 450),
  (2, 5, 'INV005', NOW(), 5, 550);
```

### invoices_has_products
```
INSERT INTO `mydb`.`invoices_has_products` (`invoices_id_invoice`, `products_id_product`, `amount`) VALUES
  (1, 1, 2),
  (1, 2, 3),
  (1, 3, 4),
  (2, 4, 5),
  (2, 5, 6),
  (2, 6, 7),
  (3, 7, 8),
  (3, 8, 9),
  (3, 9, 10),
  (4, 10, 11),
  (4, 11, 12),
  (4, 12, 13),
  (5, 13, 14),
  (5, 14, 15),
  (5, 15, 16);
```

### invoices_has_packages
```
INSERT INTO `mydb`.`invoices_has_packages` (`invoices_id_invoice`, `packages_id_package`, `amount`) VALUES
  (1, 1, 1),
  (1, 2, 1),
  (2, 2, 2),
  (2, 3, 2),
  (3, 3, 3),
  (3, 4, 3);
```

**Comment by antoniokunstek** [comment]
Dear Mislav (@mislavmatijevic),

I assigned myself to this task and I hope it's okay.
I would just like to additionally check with you the creation of a new feature with the help of git flow.
So for this issue the command would be:

git flow feature start feature/#5_Prepare_the_H2_database ?
Please check if I write the issue id correctly.

I thank you in advance for your time and wish you a pleasant rest of the day.
Best regards!
Antonio Kunštek

**Comment by mislavmatijevic** [comment]
Dear @antoniokunstek,

you don't have to use # in the branch name.
Btw, you can modify the mock INSERT INTOs and turn them into something more diverse/realistic/fun. This is pure ChatGPT output 🤷🏼 

I wish you a beautiful day. 💯 
Kind regards,
Mislav Matijević

**Comment by antoniokunstek** [comment]
Dear Mislav (@mislavmatijevic),

thank you for the ultra-fast reply.
Okay, I'll change it so that the data makes more sense.

You can expect the first pull-request soon. 😎
Best regards!
Antonio Kunštek

**Comment by mislavmatijevic** [comment]
How's it going, colleague @antoniokunstek?

**Comment by mislavmatijevic** [comment]
Hopefully, everything's ok? @antoniokunstek

**Comment by antoniokunstek** [comment]
Dear Mislav (@mislavmatijevic ),

everything is going fine. 
I was on vacation last weekend so I had limited access to the computer. I also had to do chores around the house. Sorry for the late reply. I will open a pull request for you tomorrow evening. Also, is the deal valid for this Thursday or did you go on vacation?
We can meet on discord one day to go over some details of the project, if you think it is necessary.

Once again, I am very sorry. I will make it up to you (🍻). Enjoy the rest of the week.
Best regards,
Antonio Kunštek

**Comment by antoniokunstek** [comment]
Dear Mislav (@mislavmatijevic),

I have a problem that I can't solve right now. When starting the application, my database is successfully initialized. The problem is that the 'INSERT' part of the script is constantly triggered. It is interesting that all tables, except the INVOICES table, receive more data than expected. I'm sending you screenshots below so you can see exactly what I mean. I also linked the branch to these issues, so you can check the code.

- INVOICES TABLE (expected result)

![image](https://github.com/foivz/pop-app-backend/assets/61520141/f7fa06b0-7806-48e8-b3cc-96d5b770110c)

- ROLES TABLE (unexpected result)

![image](https://github.com/foivz/pop-app-backend/assets/61520141/1291d01f-f35f-408c-a360-7f3f96089c88)

Also, I commented out the part of the code with spring-security because we will add this later so that Spring doesn't complain in the console for no reason.

I would also like to populate the database with different data, but then I realized that I don't understand the business logic very well. I was confused by the part where admin is also added for each event (according to dummy data).

I thank you in advance for your help and wish you a pleasant rest of the day.
Best regards,
Antonio Kunštek

**Comment by mislavmatijevic** [comment]
Dear @antoniokunstek.
I've fixed it for you. When you enter data manually, always use data.sql placed in resources. That, of course, does not explain why the normal INIT order commited itself 6 times (_just extra-complicated-systems things_ 🤷🏼‍♂️), but I've always done it using the `data.sql` file.

I added some more creative users in the DB, and also renamed the DB name. 😎

One small remark: the issue states in the "Then" condition of the DoD that the database is hosted on the `jdbc:h2:mem:` location. I agree that your prefered `jdbc:h2:mem:pop-dev-db` location looks much better. However I would like you to automatically update such stuff in the issue text so that I can start off testing your code without looking at such technical details in the implementation itself, rather in the issue text. 👁️

Finally, a small note - I would prefer our application to be ORM-based rather than SQL-script-based. I know I only provided the SQL script, but I would really like us to have a ORM model based on that script - not the script itself (especially since its MySQL and the application needs to be scalable). I can always take over if you can't find the time for such stupid tasks, just mention me here in the issue comment. 😃

Have a beautiful extra-stormy-armageddonish evening!

**Comment by antoniokunstek** [comment]
Dear Mislav (@mislavmatijevic),

I would like to apologize, if non-updated issue took up too much of your time. I will keep an eye on it in the future. 
I find it interesting that the file name has an effect on the execution of the script. What was most strange to me was that half of the database was working okay, and half was not (and I just could not get to the bottom of it, I wanted to understand how it works... which took too much of my time... 😒). I will use data.sql in the future, thanks for the note!

No worries, I will create model classes for ORM in the next couple of days.

Thanks for help and I want you a pleasant day.
Best regards,
Antonio Kunštek

**Comment by mislavmatijevic** [comment]
Closed via #18
